addons/website_quote/models/sale_order.py

```diff
@@ -0,0 +1,361 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import uuid
+from datetime import datetime, timedelta
+
+from odoo import api, fields, models, _
+import odoo.addons.decimal_precision as dp
+
+
+class SaleQuoteTemplate(models.Model):
+    _name = "sale.quote.template"
+    _description = "Sale Quotation Template"
+
+    name = fields.Char('Quotation Template', required=True)
+    website_description = fields.Html('Description', translate=True)
+    quote_line = fields.One2many('sale.quote.line', 'quote_id', 'Quotation Template Lines', copy=True)
+    note = fields.Text('Terms and conditions')
+    options = fields.One2many('sale.quote.option', 'template_id', 'Optional Products Lines', copy=True)
+    number_of_days = fields.Integer('Quotation Duration', help='Number of days for the validity date computation of the quotation')
+    require_payment = fields.Selection([
+        (0, 'Not mandatory on website quote validation'),
+        (1, 'Immediate after website order validation'),
+        (2, 'Immediate after website order validation and save a token'),
+    ], 'Payment', help="Require immediate payment by the customer when validating the order from the website quote")
+    mail_template_id = fields.Many2one('mail.template', 'Confirmation Mail', help="This e-mail template will be sent on confirmation. Leave empty to send nothing.")
+
+    @api.multi
+    def open_template(self):
+        self.ensure_one()
+        return {
+            'type': 'ir.actions.act_url',
+            'target': 'self',
+            'url': '/quote/template/%d' % self.id
+        }
+
+
+class SaleQuoteLine(models.Model):
+    _name = "sale.quote.line"
+    _description = "Quotation Template Lines"
+    _order = 'sequence, id'
+
+    sequence = fields.Integer('Sequence', help="Gives the sequence order when displaying a list of sale quote lines.", default=10)
+    quote_id = fields.Many2one('sale.quote.template', 'Quotation Template Reference', required=True, ondelete='cascade', index=True)
+    name = fields.Text('Description', required=True, translate=True)
+    product_id = fields.Many2one('product.product', 'Product', domain=[('sale_ok', '=', True)], required=True)
+    layout_category_id = fields.Many2one('sale.layout_category', string='Section')
+    website_description = fields.Html('Line Description', related='product_id.product_tmpl_id.quote_description', translate=True)
+    price_unit = fields.Float('Unit Price', required=True, digits=dp.get_precision('Product Price'))
+    discount = fields.Float('Discount (%)', digits=dp.get_precision('Discount'), default=0.0)
+    product_uom_qty = fields.Float('Quantity', required=True, digits=dp.get_precision('Product UoS'), default=1)
+    product_uom_id = fields.Many2one('product.uom', 'Unit of Measure ', required=True)
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        self.ensure_one()
+        product = self.product_id
```

Comment:
Use directly self.product_id in the various statements, it is easier to understand.